### PR TITLE
Fix excessive memory allocation in RawVec::reserve

### PR DESCRIPTION
**Pull request by stepancheg**
Before this patch `reserve` function allocated twice as requested
amount elements (not twice as capacity).  It leaded to unnecessary
excessive memory usage in scenarios like this:

```
let mut v = Vec::new();
v.push(17);
v.extend(0..10);
println!("{}", v.capacity());
```

`Vec` allocated 22 elements, while it could allocate just 11.

`reserve` function must have a property of keeping `push` operation
cost (which calls `reserve`) `O(1)`. To achieve this `reserve` must
exponentialy grow its capacity when it does reallocation.

There's better strategy to implement `reserve`:

```
let new_capacity = max(current_capacity * 2, requested_capacity);
```

This strategy still guarantees that capacity grows at `O(1)` with
`reserve`, and fixes the issue with `extend`.

Patch imlpements this strategy.

### CODE REVIEW

**Comment by stepancheg** [comment]
Rebased.

File: src/liballoc/raw_vec.rs [hunk 1]
```diff
@@ -350,19 +351,25 @@ impl<T> RawVec<T> {
...
+            if required_cap <= self.cap {
                 return;
             }
 
-            // Nothing we can really do about these checks :(
-            let new_cap = used_cap.checked_add(needed_extra_cap)
-                                  .and_then(|cap| cap.checked_mul(2))
-                                  .expect("capacity overflow");
+            // Cannot overflow, because cap <= isize::MAX, and type of cap is usize.
+            let double_cap = self.cap * 2;
```
Comment (bluss) [inline src/liballoc/raw_vec.rs]: This is fine, same as before, but capacity really must not exceed `isize::MAX`. I guess we used to & do rely on the allocator to OOM to save us.

I'm not sure that's sound for the element type u8 on a 32-bit platform, could we end up with a successful allocation there, larger than isize::MAX?

**Comment by bluss** [comment]
I like the new logic and can't see any obvious problems with it.

**Comment by amaranth** [comment]
If we're making changes here I think a growth factor of 1.5 would be better than 2 although I'm not sure if using jemalloc changes things here. [This blog post](https://crntaylor.wordpress.com/2011/07/15/optimal-memory-reallocation-and-the-golden-ratio/) explains the reasoning but basically with 2 you are never able to reuse an old allocation.

**Comment by stepancheg** [comment]
@amaranth grow factor is a separate, completely independent issue (I also think 1.5 would be better).

> with 2 you are never able to reuse an old allocation

Only if your application has only one `Vec`.

**Comment by Gankro** [comment]
@amaranth any static growth factor is suboptimal to my knowledge.

2 works fine for small or large vecs. 1.5 is only good for moderate-sized vecs.

https://github.com/rust-lang/rust/issues/27627

**Comment by Gankro** [comment]
Should we run some benches to make sure there aren't any issues with this? This is a no-inline cold path, so I don't think it's really necessary, but I'm not sure.

**Comment by stepancheg** [comment]
@Gankro I don't think bench is necessary, because code is straightforward, and it's similar to standard `reserve` implementations, like in [java.util.ArrayList](http://hg.openjdk.java.net/jdk7/jdk7/jdk/file/tip/src/share/classes/java/util/ArrayList.java#l183). But if you do, what a bench should look like?

> This is a ... cold path

Actually, it is not cold path, for example, when you have thousands of vectors of 2..5 elements. However, I think time spent in malloc it much more than a half of time spent in `reserve`.

**Comment by bluss** [comment]
Using `Read::read_to_end()` on an unbuffered input file is a vector reallocation benchmark.

**Comment by Gankro** [comment]
r? @bluss 

You seem to have a good handle on the issues at play here. I skimmed the code but am a bit fuzzy on if this will have the same crash behaviour when inserting into ZST collections. In particular, they rely on RawVec to crash to elide their own overflow checks for `len`.

**Comment by stepancheg** [comment]
@bluss `Read::read_to_end` is not a benchmark of `realloc`, because it does very little reallocation. Anyway, I did `io::tests::bench_read_to_end` benhmark, there's no difference:

```
test io::tests::bench_read_to_end                ... bench:  32,509,476 ns/iter (+/- 1,662,809) // master
test io::tests::bench_read_to_end                ... bench:  32,834,994 ns/iter (+/- 2,052,005) // with patch applied
```

**Comment by bluss** [comment]
@Gankro Ok, way to pass a hot potato! :smile:

From my reading, the behavior for ZST is exactly the same. Either the capacity fits and it returns, or it panics for capacity overflow.

I'll leave comments inline.

File: src/liballoc/raw_vec.rs [hunk 2]
```diff
@@ -350,19 +351,25 @@ impl<T> RawVec<T> {
 
             // NOTE: we don't early branch on ZSTs here because we want this
             // to actually catch "asking for more than usize::MAX" in that case.
-            // If we make it past the first branch then we are guaranteed to
-            // panic.
+            let required_cap = used_cap.checked_add(needed_extra_cap)
+               .expect("capacity overflow");
```
Comment (bluss) [inline src/liballoc/raw_vec.rs]: We must move this overflow check and panic case to below the `required_cap <= self.cap` check. While preserving this panic if it's a ZST element type.

The reason is that the fast path is the section of code before we decide if we need to reallocate or not, and it should be free from a branch to panic.

Comment (bluss) [inline src/liballoc/raw_vec.rs]: The first if in the original code is fine.

Comment (stepancheg) [inline src/liballoc/raw_vec.rs]: That's a good point. I overlooked it.

**Comment by stepancheg** [comment]
Updated patch with fast-path branch made first.

**Comment by bluss** [comment]
Thank you! This is beautiful.

@bors r+

**Pull request merged**